### DISCUSSION TITLE
Use fitView but position on top left

### DISCUSSION THREAD

**Discussion by artola** [xyflow/xyflow#3228] [Q&A]
When I set `fitView` it fits the flow in the viewport, but also change the position to the center/middle.
Is it possible to set the position to top-left ?

**Comment by moklick** [comment]
you need to decide if you want to fit the view (nodes are inside the viewport) or if you want to set the viewport to x=0, y=0 (that's the default). 

If you want to come up with your own fit view rules, you can use `getRectOfBounds` -> `getTransformForBounds` -> `setViewport` https://reactflow.dev/docs/api/graph-util-functions

**Reply by artola** [reply]
@moklick Thanks for the prompt answer. I do understand, and I can see that changing this `.translate(x, y)` to `.translate(0, 0)` produces the desired positioning, just it is not handy to patch the module or create 99.99% clone of the same function.
What do you think about passing `x` and `y` in the `options` of such method or like props to the component?

https://github.com/wbkd/react-flow/blob/993a778b80cc1e80a47983ed75407b579313a73c/packages/core/src/store/utils.ts#L169

This requirement arises from the fact that the appearance of the graph's center/middle positioning may not always be aesthetically pleasing due to excessive whitespace, which depends on the graph size and viewport.

**Reply by bcakmakoglu** [reply]
If you need `translate(0, 0)` why are you just *not* using `fitView`? 😄 
As mentioned by @moklick , the default *is* position 0,0 - so simply not calling `fitView` should already give you that.

Aside from that, what you're asking for (passing `x` and `y`) is already possible.

1. You can use the `defaultViewport` prop and pass the initial `x`, `y` and `zoom` values you want.

```tsx
<ReactFlow
  nodes={nodes}
  edges={edges}
  onNodesChange={onNodesChange}
  onEdgesChange={onEdgesChange}
  onConnect={onConnect}
  nodeTypes={nodeTypes}
  defaultViewport={{
    x: 100,
    y: 100,
    zoom: 1
  }}
>
  <Background />
</ReactFlow>
```

[Here's an example.](https://codesandbox.io/s/falling-glitter-798tlh?file=/src/Flow.tsx:1120-1428)

2. You can use `setViewport`.

Assuming you have wrapped your component in `ReactFlowProvider` you can do something like this.

```tsx
  const { setViewport } = useReactFlow();

  return (
    <ReactFlow
      nodes={nodes}
      edges={edges}
      onNodesChange={onNodesChange}
      onEdgesChange={onEdgesChange}
      onConnect={onConnect}
      nodeTypes={nodeTypes}
      onInit={() => {
        setViewport({
          x: 100,
          y: 100,
          zoom: 1
        });
      }}
    >
      <Background />
    </ReactFlow>
  );
```

[Here's an example.](https://codesandbox.io/s/relaxed-black-rqg4w2?file=/src/Flow.tsx:1125-1529)

**Reply by artola** [reply]
@bcakmakoglu Thanks for your help. The proposed solutions are not working as expected.

To make it easier to understand, in your example change the position of these nodes:

```
const initialNodes: Node[] = [
  {
    id: "1",
    type: "input",
    data: { label: "Node 1" },
    position: { x: 0, y: 5 }
  },
  { id: "2", data: { label: "Node 2" }, position: { x: 500, y: 100 } },
  { id: "3", data: { label: "Node 3" }, position: { x: 900, y: 100 } },
  {
    id: "4",
    type: "custom",
    data: { label: "Custom Node" },
    position: { x: 400, y: 200 }
  }
];
```
They render and do not fit the viewport, looking like this:
![image](https://github.com/wbkd/react-flow/assets/11500763/c8f1c171-16c2-4397-a11f-fe391d841c78)

With `fitView` they properly scale to fit the viewport, my problem is that they are not stick to the `defaultViewport` coordinates:
![image](https://github.com/wbkd/react-flow/assets/11500763/7be95f27-e780-4432-9624-0970c2af3eb9)

[Here a sandbox.](https://codesandbox.io/s/delicate-leftpad-mvknd8?file=/src/Flow.tsx)

IMO the positioning should happen in the initial rendering, without calling to `onInit` because it produces a lot flashing content.

**Reply by moklick** [reply]
I am still not getting the issue here. What is the exact result you want to achieve @artola ? Can you send a screenshot of the viewport you would like to start with?

**Reply by artola** [reply]
@moklick I want to use `fitView` (then content it is scaled to fit in the viewport), but the position should remain `0, 0`:

![image](https://github.com/wbkd/react-flow/assets/11500763/4c2da200-c382-42b7-9136-eeb5f1fab489)

**Reply by moklick** [reply]
This should do it then: https://codesandbox.io/s/happy-monad-wttf6p?file=/src/Flow.tsx:1649-1690 You could tweak this a bit to remove the flickering by setting opacity or something like that when you set the new viewport.

**Reply by artola** [reply]
@moklick Thanks, it works, with such flickering problem. Thus, a proper solution will be a lot better, where the only thing needed is to use the original `x` and `y` (for example, `0`) or allow some config option to "center" (as it happens already) by default in `fitViewOptions`, but possible to override.

Something around this idea: https://github.com/wbkd/react-flow/discussions/3228#discussioncomment-6467077

**Comment by artola** [comment]
@moklick If you give me some hints I would prefer to enhance `transform(x, y)` in this module or in the function that computes `x` and `y`, to use the values set on `defaultViewport` in place of the center of the vp.

https://github.com/wbkd/react-flow/blob/993a778b80cc1e80a47983ed75407b579313a73c/packages/core/src/store/utils.ts#L169

**Reply by moklick** [reply]
Something like this should work:

```js 
// your nodes bounds 
// you could use `getRectOfNodes` for this, but then you would need to set width and height for each node
const bounds = {
  x: 0,
  y: 0,
  width: 500,
  height: 500
};

const width = 1000; // width of the react flow pane
const height = 500; // height of the react flow pane

const xZoom = width / bounds.width;
const yZoom = height / bounds.height;
const zoom = Math.min(xZoom, yZoom);

const defaultViewport = { x: 0, y: 0, zoom };
```

**Reply by artola** [reply]
Why not allowing to pass into `fitViewOptions` the `x` and `y`, to use them here:

https://github.com/wbkd/react-flow/blob/993a778b80cc1e80a47983ed75407b579313a73c/packages/core/src/store/utils.ts#L169

Something like:

```
interface FitViewOptions {
  ...
  x?: number; 
  y?: number;
};


// here options contains such properties
export function fitView(get: StoreApi<ReactFlowState>['getState'], options: InternalFitViewOptions = {}) {

....

const nextTransform = zoomIdentity.translate(options.x ?? x, options.y ?? y).scale(zoom);

```

This is all needed to allow to use `fitView` and position the flow, by default as before in the middle of the viewport (backward compatible), or allow the user to set his own coordinates.

@moklick If this makes sense for you, I can PR this enhancement.

**Comment by artola** [comment]
@moklick Looking forward to it. As mentioned above, if you like the idea of enhancing the `fitViewOptions` to allow for positioning, I can submit a PR with the necessary code.

**Reply by TrueEntrant** [reply]
Hello @moklick, is there any updates on this?
